small_language_model_pretraining/nemo/pretrain_llama31.py

```diff
@@ -0,0 +1,628 @@
+    dependencies: list[str] = [],
+    retries: int = 0,
+) -> run.SlurmExecutor:
+    if not (user and host and remote_job_dir and account and partition and nodes and devices):
+        raise RuntimeError(
+            "Please set user, host, remote_job_dir, account, partition, nodes and devices args for using this function."
+        )
+
+    mounts = []
+    if custom_mounts:
+        mounts.extend(custom_mounts)
+
+    env_vars = {
+        "TRANSFORMERS_OFFLINE": "1",
+        "TORCH_NCCL_AVOID_RECORD_STREAMS": "1",
+        "NCCL_NVLS_ENABLE": "0",
+        "NVTE_DP_AMAX_REDUCE_INTERVAL": "0",
+        "NVTE_ASYNC_AMAX_REDUCTION": "1",
+        "NVTE_FUSED_ATTN": "0",
+        "TOKENIZERS_PARALLELISM": "false",
+    }
+    if custom_env_vars:
+        env_vars |= custom_env_vars
+
+    executor = run.SlurmExecutor(
+        account=account,
+        partition=partition,
+        tunnel=run.SSHTunnel(
+            user=user,
+            host=host,
+            job_dir=remote_job_dir,
+        ),
+        nodes=nodes,
+        ntasks_per_node=devices,
+        gpus_per_node=devices,
+        mem="0",
+        exclusive=True,
+        gres="gpu:8",
+        packager=run.GitArchivePackager(),
+        dependencies=dependencies,
+    )
+
+    executor.launcher = None
+    executor.container_image = container_image
+    executor.container_mounts = mounts
+    executor.env_vars = env_vars
+    executor.retries = retries
+    executor.time = time
+
+    return executor
+
+def get_pretrain(
+    size: str,
+    nnodes: int,
+    ngpus_per_node: int,
+    max_steps: int, 
+    data_module: run.Config,
+    max_lr: float=1e-4, 
+    eval_every: Optional[int]=None,
+    eval_batches: Optional[int]=None,
+) -> run.Partial:
+
+    exp_name = size
+    base_gbs = 1152
+    gbs = data_module.global_batch_size
+
+    # Providing 8B and 70B here for debugging purpose
+    # Actual benchmark should use 405B
+    if size == "8b":
+        pretrain = llm.llama3_8b.pretrain_recipe(
+            dir="/outputs",
+            name=exp_name,
+            num_nodes=nnodes,
+            num_gpus_per_node=ngpus_per_node
+        )
+
+        llama31_config = run.Config(llm.gpt.model.llama.Llama31Config8B)
+        llama31_config.seq_length = 8192
+        pretrain.model.config = llama31_config
+
+        pretrain.trainer.strategy.tensor_model_parallel_size = 1
+        pretrain.trainer.strategy.pipeline_model_parallel_size = 1
+        pretrain.trainer.strategy.virtual_pipeline_model_parallel_size = 1 # set it back to 7?
+        pretrain.trainer.strategy.context_parallel_size = 1
+
+        pretrain.optim = distributed_fused_adam_with_cosine_annealing(max_lr=3e-4)
+
+    log_lr = max_lr
+
+    # 50k samples
+    # max_steps = math.ceil(57600 * 8192 / 8192 / gbs)
+    # warmup_steps = math.ceil(57600 * 8192 / 8192 / gbs * 0.1)
+    # # 230k samples
+    max_steps = math.ceil(230000 * 8192 / 8192 / gbs)
+    warmup_steps = math.ceil(230000 * 8192 / 8192 / gbs * 0.1)
```

Comment:
max_steps should be fixed at 1.2M
warmup_steps should be parametrizable from the config